sdk/communication/azure-communication-phonenumbers/azure/communication/phonenumbers/siprouting/_sip_routing_client.py

```diff
@@ -0,0 +1,259 @@
+
+        :param trunk: Trunk object to be set.
+        :type trunk: ~azure.communication.siprouting.models.SipTrunk
+        :returns: None
+        :rtype: None
+        :raises: ~azure.core.exceptions.HttpResponseError, ValueError
+        """
+        if trunk is None:
+            raise ValueError("Parameter 'trunk' must not be None.")
+
+        self._patch_trunks_([trunk],**kwargs)
+
+    @distributed_trace
+    def delete_trunk(
+        self,
+        trunk_fqdn,  # type: str
+        **kwargs  # type: Any
+    ):  # type: (...) -> None
+        """Deletes SIP trunk.
+
+        :param trunk_fqdn: FQDN of the trunk to be deleted.
+        :type trunk_fqdn: str
+        :returns: None
+        :rtype: None
+        :raises: ~azure.core.exceptions.HttpResponseError, ValueError
+        """
+        if trunk_fqdn is None:
+            raise ValueError("Parameter 'trunk_fqdn' must not be None.")
+
+        self._rest_service.patch_sip_configuration(
+            body=SipConfiguration(trunks={trunk_fqdn:None}),
+            **kwargs)
+
+    @distributed_trace
+    def get_trunks(
+        self,
+        **kwargs  # type: Any
+    ):  # type: (...) -> Iterable[SipTrunk]
+        """Retrieves an iterable of currently configured SIP trunks.
+
+        :returns: Current SIP trunks configuration.
+        :rtype: Iterable[~azure.communication.siprouting.models.SipTrunk]
+        :raises: ~azure.core.exceptions.HttpResponseError
+        """
+        return self._get_trunks_(**kwargs)
+
+    @distributed_trace
+    def get_routes(
+        self,
+        **kwargs  # type: Any
+    ):  # type: (...) -> Iterable[SipTrunkRoute]
+        """Retrieves an iterable of currently configured SIP routes.
+
+        :returns: Current SIP routes configuration.
+        :rtype: Iterable[~azure.communication.siprouting.models.SipTrunkRoute]
+        :raises: ~azure.core.exceptions.HttpResponseError
+        """
+        config = self._rest_service.get_sip_configuration(
+            **kwargs
+        )
+        return config.routes
+
+    @distributed_trace
+    def set_trunks(
+        self,
+        trunks,  # type: List[SipTrunk]
+        **kwargs  # type: Any
+    ):  # type: (...) -> None
+        """Overwrites the list of SIP trunks.
+
+        :param trunks: New list of trunks to be set.
+        :type trunks: List[SipTrunk]
+        :returns: None
+        :rtype: None
+        :raises: ~azure.core.exceptions.HttpResponseError, ValueError
+        """
+        if trunks is None:
+            raise ValueError("Parameter 'trunks' must not be None.")
+
+        old_trunks = self._get_trunks_(**kwargs)
+
+        if len(old_trunks) > 0:
+            self._rest_service.patch_sip_configuration(
+                body=SipConfiguration(trunks={x.fqdn: None for x in old_trunks}), **kwargs
+            )
+        if len(trunks) > 0:
+            self._patch_trunks_(trunks, **kwargs)
```

Comment:
Going multiple REST calls within a single API can be problematic.
Are lines 207 and 211 mutually exclusive? Only one of these two will be called right?
Is this method idempotent if a network failure for example interrupted the patch operation and it was retried?

Is this pattern of GET -> PATCH the same pattern that's used across languages?

Reply:
Changed the call logic, to have just one GET and one PATCH.

This pattern is used across languages, we dont have another option to wipe the trunks other than first retrieving them and then deleting. From the point of view of idempotency, this should be ok.
The operation will always retrieve the current list of trunks and overwrite them, so idempotency shouldn't be an issue.

An issue could be concurrency, but as discussed at the Arch reviews, it doesn't make sense from user perspective to update the configuration from multiple places at once, because it could always cause inconsistent configuration, so concurrency is not a concern.